### ISSUE TITLE
Suggestion: Separate file for notebook executed cell outputs.

### ISSUE THREAD

**Issue by jbursey** [jupyter/notebook#5677]
Unless this is a feature already I think it would be nice to have a separate file (something like .ipynb.output) that links output to their cells in the .ipynb json file. This would make it significantly easier to exclude notebook outputs in source control systems like git.

If this is already possible somehow I would be interested to know.

**Comment by gitjeff05** [comment]
Its not a bad idea. But if keeping cell output out of source control is your primary concern, the easiest solution is to just clear the outputs before committing. There are a few ways to do that:

1. Use a [commit hook as outlined in Jupyter docs](https://jupyter-notebook.readthedocs.io/en/stable/extending/savehooks.html).

2. Use Jupyter's shortcut to "clear all cell output"

3. Use [nbconvert to clear the notebook outputs](https://nbconvert.readthedocs.io/en/latest/api/preprocessors.html#nbconvert.preprocessors.ClearOutputPreprocessor) before committing.

4. You could also just write your own shell script to clear outputs. [I wrote one using `jq` to do that and it is fairly easy](https://github.com/gitjeff05/kapitsa/blob/master/lib/remove_cell_outputs.sh#L66).

Some folks also choose to just convert the notebook to python using nbconvert and then just commit that. If you search for "How to version control jupyter notebooks" you will see a bunch of posts on the topic.

**Comment by cipri-tom** [comment]
I think that jupyterlab already has the capability of displaying the output in a different view from the notebook.

**Comment by IvoMerchiers** [comment]
Alternatively, [Jupytext](https://github.com/mwouts/jupytext) could be helpful for your case. It allows you to save notebooks as code. Then you only need to commit the code to git, whilst you can ignore the notebooks for version control. 

Their paired notebooks avoid the need for automatically saving and converting the notebooks.

**Comment by starball5** [comment]
Related: jupyterlab/jupyterlab#9444 and jupyterlab/jupyterlab-git#392

Related question on Stack Overflow: [How can I configure my tools to ignore or prevent updates to the execution_count field in a Jupyter Notebook from being tracked in git?](https://stackoverflow.com/q/75575346/11107541)

**Comment by th0ger** [comment]
Good idea. The alternative discussed above are about excluding cells from source control.

But sometimes we have a need to include the executed cells in source control. (My current case is with Quarto.)
Including the cell output in the .ipynb file makes it extremely difficult to review/diff a plaintext. This experience would be improved a lot if the input and output could be separated. A reviewer would then be able to decide whether the changes was cause by code change, or purely external changes and rejection of the notebook.

**Comment by th0ger** [comment]
@alexbjorling LFS is a good point. Notebook output is very suitable for LFS, but input cells are not.

**Comment by Tyrrx** [comment]
I think cleaning the notebook can only be seen as a workaround.

**Comment by zmbc** [comment]
Yes, this would be a huge improvement. I believe this is why Quarto [embeds Python in Markdown](https://quarto.org/docs/computations/python.html) as a "plain text representation of notebooks."

If the .ipynb itself could be in a readable plain-text format, and the outputs stored in a separate file, that would:
- Make diffing of notebook code trivial.
- Make editing (the code of) a notebook easy using any text editor.
- Allow the output to be versioned using a non-plaintext scheme, e.g. Git LFS as mentioned above, or being snapshotted only periodically as opposed to on every commit.

**Comment by carschandler** [comment]
Hugely in support of this! Even if it isn't a default behavior, it would be amazing to have the option.

**Comment by zmbc** [comment]
Surprised not to see anyone mention this yet, this jupyter extension does almost exactly what this thread describes: https://jupytext.readthedocs.io/en/latest/paired-notebooks.html

**Comment by lukehutch** [comment]
There's a new compelling reason to do this. Claude Code works really well with ipynb, e.g. to hammer out a proof, except that once you start running the cells, you quickly exceed the context window size, and Claude Code can't reason about the notebook anymore in its entirety, it can only grep through the notebook for keywords, and store summaries in its own session context:

> Error: File content (30950 tokens) exceeds maximum allowed tokens (25000). Please use offset and limit parameters to read specific portions of the file, or use the GrepTool to search for specific content.

Cleaning the notebook on git commit doesn't solve this problem when you are in between commits and need to run cells, and then keep working with Claude.

**Comment by jtpio** [comment]
@lukehutch you may be interested in `jupyter-server-documents`, iirc it deals with outputs a bit differently mostly for this AI agent use case: https://github.com/jupyter-ai-contrib/jupyter-server-documents
